xprof/tally.cpp

```diff
@@ -32,14 +38,46 @@ void print_metadata(std::vector<std::string> metadata) {
     std::cout << value << std::endl;
 }
 
+int get_backend_id(std::string name){
+  for(int i = 0; i < BACKEND_MAX; ++i)
+    // backend_name is located in xprof_utils.cpp
+    if (std::string{backend_name[i]} == name) return i;
+  return -1;
+}
+
 void btx_initialize_usr_data(void *btx_handle, void **usr_data) {
   /* User allocates its own data structure */
-  *usr_data = new tally_dispatch_t;
+  auto *data = new tally_dispatch_t;
+  *usr_data = data;
+
+  /* Backend information must match enum backend_e in xprof_utils.hpp */
+  data->backend_level = {
+    2, // BACKEND_UNKNOWN
+    2, // BACKEND_ZE
+    2, // BACKEND_OPENCL
+    2, // BACKEND_CUDA
+    1, // BACKEND_OMP_TARGET_OPERATIONS
+    0, // BACKEND_OMP
+    2  // BACKEND_HIP
+  };
 }
 
 void btx_read_params(void *btx_handle, void *usr_data, btx_params_t *usr_params) {
   tally_dispatch_t *data = (tally_dispatch_t *)usr_data;
   data->params = usr_params;
+
+  // Consumes key:value pairs in the stringstream k1:v1,..,kn:vn
+  std::stringstream tokens{data->params->backend_level};
+  std::string tmp;
+  while (getline(tokens, tmp, ',')) {
+    std::stringstream tmp_string{tmp};
+    std::string k,v;
+    getline(tmp_string, k, ':');
+    int id = get_backend_id(k);
+    assert((id > 0) && "Backend not found. Please check --backend-level format.");
```

Comment:
I'd rather have this return an error code than use assert here. The error code can then be propagated to the babeltrace level which will make the component initialization fail gracefully.

If I am in ruby trying to initialize a plugin and pass it invalid argument, the plugin should fail gracefully, not terminate my ruby virtual machine.

I understand that this will require updates to metababel, but this will need to be fixed ASAP.

Reply:
How should we down a plugin gracefully and returning a nice error code? Sending some error code in the `consuming next`?

Reply:
In initialization, this is easy, `sink_initialize` in `btx_main.c` should return `BT_COMPONENT_CLASS_INITIALIZE_METHOD_STATUS_ERROR` if `btx_initialize_usr_data` returns an error status (could be a boolean set to `false` for instance), or `BT_COMPONENT_CLASS_INITIALIZE_METHOD_STATUS_MEMORY_ERROR` in case one of the `calloc` fails.

In the more general sense, it is very similar in the other part of the API where each babeltrace callback can return `STATUS_ERROR` codes.

Reply:
This seemed to be a non-critical program option. We can do something like, "Backend not found. Default value used..." and continue the execution or just make the program fail.

I am not sure if I understood well, babeltrace is a good idea, but I think we will miss the ability to track the error here, and the babeltrace error reporting  is not as good as it should do. We should have an error code for every possible error in order to track that. Correct me if I am wrong.

Reply:
BT2 has pretty nice error reporting features, see https://babeltrace.org/docs/v2.0/libbabeltrace2/group__api-error.html
You can log where and why an error was raised and these get collected in a stack.

Anyway, I don't think it can be fixed in this PR since it requires a new meta-babel that would be "error aware", but I raised it here because the user can make everything crash very ungracefully just by specifying a malformed option on the command line (and the solution in this case is pretty easy to envision).